src/txrebroadcast.cpp

```diff
@@ -0,0 +1,242 @@
+            continue;
+        } else if (entry_it->m_last_attempt > start_time - MIN_REATTEMPT_INTERVAL) {
+            // We already rebroadcasted this in the past 4 hours. Even if we
+            // added it to the set, it would probably not get INVed to most
+            // peers due to filterInventoryKnown.
+            continue;
+        } else {
+            // We have rebroadcasted this transaction before, but will try
+            // again now. Record the attempt.
+            auto UpdateRebroadcastEntry = [start_time](RebroadcastEntry& rebroadcast_entry) {
+                rebroadcast_entry.m_last_attempt = start_time;
+                ++rebroadcast_entry.m_count;
+            };
+
+            m_attempt_tracker->modify(entry_it, UpdateRebroadcastEntry);
+        }
+
+        // Add to set of rebroadcast candidates
+        rebroadcast_txs.push_back(TxIds(txid, wtxid));
+    }
+
+    TrimMaxRebroadcast();
+
+    auto delta1 = after_cnb_time - start_time;
+    auto delta2 = GetTime<std::chrono::microseconds>() - start_time;
+    LogPrint(BCLog::BENCH, "GetRebroadcastTransactions(): %d us total, %d us spent in CreateNewBlock.\n", delta2.count(), delta1.count());
+    LogPrint(BCLog::NET, "Queued %d transactions for attempted rebroadcast, filtered from %d candidates with cached fee rate of %s.\n", rebroadcast_txs.size(), block_template->block.vtx.size() - 1, m_cached_fee_rate.ToString(FeeEstimateMode::SAT_VB));
+
+    for (TxIds ids : rebroadcast_txs) {
+        LogPrint(BCLog::NET, "Attempting to rebroadcast txid: %s, wtxid: %s\n", ids.m_txid.ToString(), ids.m_wtxid.ToString());
+    }
+
+    return rebroadcast_txs;
+};
+
+void TxRebroadcastHandler::CacheMinRebroadcastFee()
+{
+    if (m_chainman.ActiveChainstate().IsInitialBlockDownload()) return;
+
+    LOCK(m_rebroadcast_mutex);
+
+    // Update stamp of chain tip on cache run
+    m_tip_at_cache_time = m_chainman.ActiveTip();
+
+    // Store the existing fee rate
+    m_previous_cached_fee_rate = m_cached_fee_rate;
+
+    // Calculate a new cached fee rate
+    auto start_time = GetTime<std::chrono::microseconds>();
+    m_cached_fee_rate = BlockAssembler(m_chainman.ActiveChainstate(), m_mempool, m_chainparams).MinTxFeeRate();
+    auto delta_time = GetTime<std::chrono::microseconds>() - start_time;
+    LogPrint(BCLog::BENCH, "Caching minimum fee for rebroadcast to %s, took %d us to calculate.\n", m_cached_fee_rate.ToString(FeeEstimateMode::SAT_VB), delta_time.count());
+};
+
+void TxRebroadcastHandler::RemoveFromAttemptTracker(const CTransactionRef& tx) {
+    LOCK(m_rebroadcast_mutex);
+    const auto it = m_attempt_tracker->find(tx->GetWitnessHash());
+    if (it == m_attempt_tracker->end()) return;
+    m_attempt_tracker->erase(it);
+}
+
+void TxRebroadcastHandler::TrimMaxRebroadcast()
+{
+    // Delete any entries that are older than MAX_ENTRY_AGE
+    auto min_age = GetTime<std::chrono::microseconds>() - MAX_ENTRY_AGE;
+
+    while (!m_attempt_tracker->empty()) {
```

Comment:
The two while loops can be consolidated:

```diff
diff --git a/src/txrebroadcast.cpp b/src/txrebroadcast.cpp
index 940735df1a..401621153c 100644
--- a/src/txrebroadcast.cpp
+++ b/src/txrebroadcast.cpp
@@ -195,18 +195,11 @@ void TxRebroadcastHandler::TrimMaxRebroadcast()
 
     while (!m_attempt_tracker->empty()) {
         auto it = m_attempt_tracker->get<index_by_last_attempt>().begin();
-        if (it->m_last_attempt < min_age) {
-            m_attempt_tracker->get<index_by_last_attempt>().erase(it);
-        } else {
+        if (it->m_last_attempt >= min_age && m_attempt_tracker->size() <= MAX_ENTRIES) {
+            // The transaction is not old enough to trim and the attempt tracker is not full
             break;
         }
     }
-
-    // If there are still too many entries, delete the oldest ones
-    while (m_attempt_tracker->size() > MAX_ENTRIES) {
-        auto it = m_attempt_tracker->get<index_by_last_attempt>().begin();
-        m_attempt_tracker->get<index_by_last_attempt>().erase(it);
-    }
 };
```

Perhaps that's easier to read since there's less repetition?

If you do that, then perhaps you could just remove this function entirely and put the loop inside `GetRebroadcastTransactions()` since it's only called in that one place.

Reply:
good point on consolidating the two loops. the code you shared doesn't quite make sense (it no longer erases anything from the attempt tracker? 😛), but I got the gist and reworked the logic.